doc/sphinx/proofs/writing-proofs/proof-mode.rst

```diff
@@ -1,74 +1,176 @@
----------------------------------------
+Entering and exiting proof mode
+-------------------------------
+
+Coq enters :term:`proof mode` when you begin a proof through
+commands such as :cmd:`Theorem` or :cmd:`Goal`.  Coq user interfaces
+usually have a way to indicate that you're in proof mode.
+
+:term:`Tactics <tactic>` are available only in proof mode (currently giving syntax
+error outside of proof mode).  Most :term:`commands <command>` can be used both in and out of
+proof mode, but some commands only work in proof mode.
+
+.. todo PR: following seems to conflate the proof context with section-level info, e.g.
+   Parameter, ... are part of the section  Is this truly the "local context"?
 
-The proof editing mode is entered by asserting a statement, which typically is
-the assertion of a theorem using an assertion command like :cmd:`Theorem`. The
-list of assertion commands is given in :ref:`Assertions`. The command
-:cmd:`Goal` can also be used.
+   To each goal is associated a number of hypotheses called the *local context*
+   of the goal. Initially, the local context contains the local variables and
+   hypotheses of the current section (see Section :ref:`gallina-assumptions`) and
+   the local variables and hypotheses of the theorem statement.
+
+When the proof is completed, you can exit proof mode with commands such as
+:cmd:`Qed`, :cmd:`Defined` and :cmd:`Save`.
 
 .. cmd:: Goal @type
 
-   This is intended for quick assertion of statements, without knowing in
-   advance which name to give to the assertion, typically for quick
-   testing of the provability of a statement. If the proof of the
-   statement is eventually completed and validated, the statement is then
-   bound to the name ``Unnamed_thm`` (or a variant of this name not already
-   used for another statement).
+   Asserts an unnamed proposition.  This is intended for quick tests that
+   a proposition is provable.  If the proof is eventually completed and
+   validated, you can assign a name with the :cmd:`Save` or :cmd:`Defined`
+   commands.  If no name is given, the name will be `Unnamed_thm` (or,
+   if that name is already defined, a variant of that).
 
 .. cmd:: Qed
 
-   This command is available in interactive editing proof mode when the
-   proof is completed. Then :cmd:`Qed` extracts a proof term from the proof
-   script, switches back to Coq top-level and attaches the extracted
-   proof term to the declared name of the original goal. The name is
-   added to the environment as an opaque constant.
+   Passes a completed proof term to Coq's kernel to verify that the type of
+   proof term matches the theorem statement.  If it's verified, the
+   proof term is added to the global context as an opaque constant
+   using the declared name from the original goal.
+
+   It's very rare for a proof term to fail verification.  Generally this
+   indicates a bug in Coq and/or a third party plugin used to
+   create the proof term.
```

Comment:
I would rather say that "generally this indicates a bug in a tactic you used or that you misused some unsafe tactics".